### DISCUSSION TITLE
UniqueConstraintException: Parse out index name and duplicate values

### DISCUSSION THREAD

**Discussion by EricHarmon** [Giorgi/EntityFramework.Exceptions#70] [Ideas]
I just started implementing this package today, to replace some home-grown code that I have. I sometimes handle unique constraint exceptions differently in my code; for example, a duplicate on IX_JobNumber might display a message such as "That job number is in use. You must supply a unique job number, or leave the job number blank". A duplicate on IX_SKU (for the same table) might display a message such as "Please enter a unique SKU".

It would be useful to me if this library would parse the error message appropriately for each supported database, and provide the index name as a property string ConstraintName, and perhaps the list of values as a property List<string> ConstraintValues.

**Comment by Giorgi** [comment]
I have thought about it but I don't know if the constraint name is included in the message on every database. Also, the error message can be localized by the database so I would need to support all the languages supported by the database server for it to work properly.

**Comment by EricHarmon** [comment]
Understood. I knew it was more complicated than I require for my own needs.

I wonder if the ConstraintName could simply be left at null if the database system doesn’t support it? Of course, the error message is likely a larger issue. I’m fortunate in that all my clients are English-speaking so I haven’t worried about multiple languages in my own code.

I guess for my use, I can create MyUniqueConstraintException, and then in my DbContext derivative, in very loose pseudocode, I can simply do this:

catch(UniqueConstraintException ex)
{
  // Parse the index name and key values in English
  throw new MyUniqueConstraintException(IndexName, KeyValues, ex);
}

That’s not difficult for me to do within my SQLServer and English-only environment, so I’ll just go that route.

Thank you for listening. You can feel free to close this discussion if you like 😊

**Comment by Giorgi** [comment]
This is now implemented in the latest commit. Thanks to @NickStrupat for the idea

**Comment by Giorgi** [comment]
This is now included in the 8.1.0 version of the library. See ReadMe for more details.

**Comment by EricHarmon** [comment]
This is great! It allowed me to simplify my code some. I'm doing a similar thing with ReferenceConstraintException, where I'm catching it, parsing out the reference name, the referenced table name, and the referenced column name, and then rethrowing the exception. I'd like to request an addition to the ReferenceConstraintException, where it will do the parsing for me as well.

**Reply by Giorgi** [reply]
Implemented in https://github.com/Giorgi/EntityFramework.Exceptions/commit/53da32c144860652e8f3bcad717051560ea6c670

By the way, the project accepts sponsorship and you could sponsor it to show your appreciation.
